Move list field definition doc from SonataDoctrineORMAdminBundle

eerison opened the pull request:
<!-- THE PR TEMPLATE IS NOT AN OPTION. DO NOT DELETE IT, MAKE SURE YOU READ AND EDIT IT! -->
## Moved list field definition from SonataDoctrineORMAdminBundle

<!-- Describe your Pull Request content here -->

<!--
    Show us you choose the right branch.
    Different branches are used for different things :
    - 4.x is for everything backwards compatible, like patches, features and deprecation notices
    - 5.x is for deprecation removals and other changes that cannot be done without a BC-break
    More details here: https://github.com/sonata-project/SonataAdminBundle/blob/4.x/CONTRIBUTING.md#base-branch
-->

<!--
    Specify which issues will be fixed/closed.
    Remove it if this is not related.
-->

## Changelog

No changelog for doc

Issue: https://github.com/sonata-project/SonataDoctrineORMAdminBundle/issues/1613

eerison committed:
Feature/moved list field definition from SonataDoctrineORMAdminBundle

eerison commented:
hmmm any idea why `PHPstan` has failed in files that I didn't change 😕

VincentLanglet commented:
> hmmm any idea why `PHPstan` has failed in files that I didn't change 😕

Don't worry, the build is failing on the stable branch ; https://github.com/sonata-project/SonataAdminBundle/pull/7701 will fix it.

But I had in mind in https://github.com/sonata-project/SonataDoctrineORMAdminBundle/issues/1613#issuecomment-1009970338 to move more doc to SonataAdmin.

Don't hesitate to move the whole https://docs.sonata-project.org/projects/SonataDoctrineORMAdminBundle/en/4.x/reference/list_field_definition/#list-actions 4.3 section for instance. And the whole 4.4 could be moved too.

VincentLanglet commented on docs/reference/action_list.rst:
```diff
@@ -153,7 +157,7 @@ Options
 - ``type`` (m): defines the field type - mandatory for the field description
   itself but will try to detect the type automatically if not specified
-- ``template`` (o): the template used to render the field
+- ``template`` (o): the template used to render the field, Your twig file need to be saved in **templates** folder, for example **Admin/list__action_[YOUR_ACTION].html.twig**
```
This message is too specific to the actions. But this options is used for field templates too. https://docs.sonata-project.org/projects/SonataDoctrineORMAdminBundle/en/4.x/reference/list_field_definition/#custom-template

eerison replied:
Removed ;)

eerison commented:
Hey @VincentLanglet 

I moved the sections 4.3 and 4.4

But for 4.3 I didn't moved

```
You can set actions for the list items by adding an ‘_action’ field in configureListFields:

$listMapper
    ->add(ListMapper::NAME_ACTIONS, ListMapper::TYPE_ACTIONS, []
        'actions' => [
            'show' => [],
            'edit' => [],
        ]
    ]);
```

IMO it's the same thing that already exist in the documentation, isn't it?


and `4.4.3. Custom route` I moved to `Routes` section.

VincentLanglet commented on docs/reference/action_list.rst:
```diff
@@ -801,3 +847,76 @@ Combine this with configuring a custom template and you'll have a list column fu
+Advance Usage
+-------------
+
+Displaying sub entity properties
```
In the 
```
**TODO**:
    * adding custom columns
    * targeting submodel fields using dot-separated notation
```
at the top of the doc, you can now remove 
```
targeting submodel fields using dot-separated notation
```

VincentLanglet replied:
And I would say the custom columns could be remove too, since we provide example with templates and there is "Displaying a non-model field" examples

eerison replied:
`Displaying a non-model field` Removed

eerison replied:
> In the
> 
> ```
> **TODO**:
>     * adding custom columns
>     * targeting submodel fields using dot-separated notation
> ```
> 
> at the top of the doc, you can now remove
> 
> ```
> targeting submodel fields using dot-separated notation
> ```

removed

VincentLanglet replied:
Can you remove also
```
**TODO**:
    * adding custom columns
```
To me, it's done. Custom columns are made with custom template. Or am I missing something ?

VincentLanglet commented on docs/reference/action_list.rst:
```diff
@@ -186,6 +217,21 @@ Available types and associated options
 | ``FieldDescriptionInterface::TYPE_*`` |                     | See :doc:`Field Types <field_types>`                                  |
 +---------------------------------------+---------------------+-----------------------------------------------------------------------+
 
+Setting up custom template
```
We already talk about custom template somewhere else, and here we talk only about actions.

I would use a title like `Setting up custom action buttons`

eerison replied:
Done!

VincentLanglet commented on docs/reference/action_list.rst:
```diff
@@ -186,6 +217,21 @@ Available types and associated options
+Setting up custom template
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+You can specify your own by setting up the 'template' option like so::
```
```suggestion
You can specify your own action buttons by setting up the 'template' option like so::
```

VincentLanglet commented on docs/reference/action_list.rst:
```diff
@@ -732,72 +777,77 @@ Checkbox range selection
-https://docs.sonata-project.org/projects/SonataAdminBundle/en/4.x/cookbook/recipe_virtual_field/
-
-Combine this with configuring a custom template and you'll have a list column fully customizable in what it eventually renders. ::
+.. code-block:: jinja
```
```suggestion
.. code-block:: twig
```

Shouldn't it be twig instead ?

eerison replied:
Done

VincentLanglet commented on docs/reference/action_list.rst:
```diff
@@ -732,72 +772,77 @@ Checkbox range selection
     You can check / uncheck a range of checkboxes by clicking a first one,
     then a second one with shift + click.
 
-Displaying a non-model field
```
Just saw that, I think you missunderstand me.

I said that because we have the `displaying a non model field` section, the 
```
**TODO**: * adding custom columns
```
could be removed.

So I wanted to keep all this section ^^'

eerison replied:
Ohhh OK ;)

eerison replied:
Added again.

franmomu commented on docs/reference/action_list.rst:
```diff
@@ -801,3 +841,76 @@ Combine this with configuring a custom template and you'll have a list column fu
+        {
+            $list
+                ->addIdentifier('id')
+                ->add('image', 'string', ['template' => '@SonataMedia/MediaAdmin/list_image.html.twig'])
+                ->add('custom', 'string', ['template' => '@SonataMedia/MediaAdmin/list_custom.html.twig'])
```
Should we use here `FieldDescriptionInterface::TYPE_STRING` constant? instead of `string`

eerison replied:
Yes,  it's used above!

Thank you @franmomu

franmomu commented on docs/reference/action_list.rst:
```diff
@@ -801,3 +841,77 @@ Combine this with configuring a custom template and you'll have a list column fu
+Advance Usage
+-------------
+
+Displaying sub entity properties
```
Since this is `SonataAdminBundle` I would use `model` instead of `entity` in general.

eerison replied:
Done

franmomu commented on docs/reference/action_list.rst:
```diff
@@ -801,3 +841,77 @@ Combine this with configuring a custom template and you'll have a list column fu
+        }
+    }
+
+.. note::
+
+    This only makes sense when the prefix path is made of entities, not collections.
+
```
I think I would add this at the beginning with the first sentence (just a personal suggestion, feel free to ignore if you think otherwise)

eerison replied:
Hey @franmomu I putted this bellow of 

```
Displaying sub model properties
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

please look if it's ok!

eerison replied:
<img width="724" alt="Screenshot 2022-01-17 at 13 33 44" src="https://user-images.githubusercontent.com/6358755/149807487-d4352622-2789-44b1-af18-4582f2824c74.png">

eerison replied:
<img width="724" alt="Screenshot 2022-01-17 at 13 33 44" src="https://user-images.githubusercontent.com/6358755/149807645-198444e4-8e84-4594-ac42-9df4ee74d3de.png">

franmomu commented on docs/reference/action_list.rst:
```diff
@@ -801,3 +841,77 @@ Combine this with configuring a custom template and you'll have a list column fu
+        protected function configureListFields(ListMapper $list): void
+        {
+            $list
+                ->addIdentifier('id')
+                ->addIdentifier('firstName')
+                ->addIdentifier('lastName')
+                ->addIdentifier('address.street')
+                ->addIdentifier('address.ZIPCode')
+                ->addIdentifier('address.town')
```
```suggestion
                ->addIdentifier('id')
                ->add('firstName')
                ->add('lastName')
                ->add('address.street')
                ->add('address.ZIPCode')
                ->add('address.town')
```

eerison replied:
Done

franmomu commented on docs/reference/action_list.rst:
```diff
@@ -50,6 +44,33 @@ For more detailed information about routes, see :doc:`routing`::
+Default route for a link is `show` (for `FieldDescriptionInterface::TYPE_MANY_TO_ONE` and `FieldDescriptionInterface::TYPE_ONE_TO_ONE`).
+Using this, the route can be customized as follows::
+
+    namespace Sonata\MediaBundle\Admin;
```
```suggestion
    namespace App\Admin;
```
If these are examples for users to use, I would use `App` namespace (in all examples) to make it clear.

eerison replied:
Done!

VincentLanglet commented:
@eerison If you rebase 4.x the lint build will be fixed

eerison commented:
> @eerison If you rebase 4.x the lint build will be fixed

Nice, I'll check the comments today and rebase the branch ;)

franmomu commented on docs/reference/action_list.rst:
```diff
@@ -801,3 +841,76 @@ Combine this with configuring a custom template and you'll have a list column fu
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+.. note::
+
+    This only makes sense when the prefix path is made of entities, not collections.
```
```suggestion
    This only makes sense when the prefix path is made of models, not collections.
```

franmomu commented on docs/reference/action_list.rst:
```diff
@@ -801,3 +841,76 @@ Combine this with configuring a custom template and you'll have a list column fu
+
+If you need a specific layout for a row cell, you can define a custom template::
+
+    namespace Sonata\MediaBundle\Admin;
```
```suggestion
    namespace App\Admin;
```

eerison committed:
Update docs/reference/action_list.rst

Co-authored-by: Fran Moreno <franmomu@gmail.com>

eerison committed:
Update docs/reference/action_list.rst

Co-authored-by: Fran Moreno <franmomu@gmail.com>

franmomu commented on docs/reference/action_list.rst:
```diff
@@ -801,3 +841,76 @@ Combine this with configuring a custom template and you'll have a list column fu
+Advance Usage
+-------------
+
+Displaying sub model properties
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
```suggestion
Displaying sub model properties
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Sorry I've just saw this

eerison committed:
Update docs/reference/action_list.rst

Co-authored-by: Fran Moreno <franmomu@gmail.com>

VincentLanglet commented:
Thanks
